filter.go

```diff
@@ -0,0 +1,280 @@
+	ErrEmptyAttrValue   = errors.New("empty attribute value")
+	ErrMissingOperation = errors.New("missing filter operation")
+)
+
+// FilterNode represents an individual filter element of a parsed LDAP filter
+// string. A full filter is an abstract syntax tree (AST) made of FilterNodes.
+//
+// LDAP filters: https://ldap.com/ldap-filters/
+type FilterNode interface {
+	Match(e *Entry) bool
+}
+
+// Parse parses an LDAP filter string into an [FilterNode] AST that can be
+// used to match against an LDAP Entry. If the filter string could not be
+// parsed due to a syntax error or unimplemented filter, an error is
+// returned with a nil FilterNode.
+func Parse(filter string) (n FilterNode, err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			if perr, ok := r.(error); ok {
+				n, err = nil, perr
+			}
+		}
+	}()
+
+	filter = strings.TrimSpace(filter)
+	c := &cursor{input: []rune(filter)}
+	n = parseFilter(c)
+
+	if !c.isEOF() {
+		panicf("%w: %s", ErrUnexpectedInput, string(c.input[c.pos:]))
+	}
+
+	return n, nil
+}
+
+func parseFilter(c *cursor) FilterNode {
+	c.expectRune('(')
+	switch c.peek() {
+	case '&':
+		return parseAndOrFilter(c, '&')
+	case '|':
+		return parseAndOrFilter(c, '|')
+	case '!':
+		return parseNotFilter(c)
+	default:
+		return parseOpFilter(c)
+	}
+}
+
+func parseAndOrFilter(c *cursor, op rune) FilterNode {
+	var nodes []FilterNode
+	c.expectRune(op)
+
+	for {
+		if c.peek() == '(' {
+			nodes = append(nodes, parseFilter(c))
+		} else {
+			c.expectRune(')')
+			break
+		}
+	}
+
+	if len(nodes) == 0 {
+		panicf("%w: expected filter, got ')'", ErrUnexpectedInput)
+	}
+
+	switch op {
+	case '&':
+		return &And{Nodes: nodes}
+	case '|':
+		return &Or{Nodes: nodes}
+	}
+	panicf("%w: unknown and/or op: %q", ErrInternal, op)
+	// NOTREACHED
+	return nil
+}
+
+func parseNotFilter(c *cursor) FilterNode {
+	c.expectRune('!')
+	node := parseFilter(c)
+	c.expectRune(')')
+	return &Not{Node: node}
+}
+
+func parseOpFilter(c *cursor) FilterNode {
+	rs := c.extractTo(')')
+	c.expectRune(')')
+
+	var attr, op, value string
+	switch idx := slices.Index(rs, '='); {
+	case idx == -1:
+		panic(ErrMissingOperation)
+	case idx > 0 && isOp(string(rs[idx-1:idx+1])):
+		// 2-char op with second char being =
+		attr = validateAttrName(rs[:idx-1])
+		op = string(rs[idx-1 : idx+1])
+		value = string(rs[idx+1:])
+	default:
+		// Op is '='
+		attr = validateAttrName(rs[:idx])
+		op = string(rs[idx])
+		value = string(rs[idx+1:])
+	}
+
+	if value == "" {
+		panic(ErrEmptyAttrValue)
+	}
+
+	if op == "=" && value == "*" {
+		return &Presence{Attr: attr}
+	}
+	if op == "=" {
+		// TODO: Implement Substring filters (value contains '*' globs)
+		return &Equality{Attr: attr, Value: value}
+	}
+
+	// TODO: Implement >= (Greater-Or-Equal) and <= (Less-Or-Equal) filters.
+	// TODO: Implement ~= (Approximate Match) filter (maybe)
+	panicf("%w: operation: %s", ErrUnimplemented, op)
+	return nil
+}
+
+func validateAttrName(rs []rune) string {
```

Comment:
shouldn't this `validateAttrNames` function also be called in other places, like when loading from JSON or search request attributes or DN parsing?